subsys/net/lib/nrf_cloud/src/nrf_cloud_pgps_utils.c

```diff
@@ -0,0 +1,504 @@
+{
+	int i;
+
+	LOG_DBG("resetting block pool");
+	pool.first_free = 0;
+	for (i = 0; i < NUM_BLOCKS; i++) {
+		pool.block_used[i] = false;
+	}
+}
+
+void npgps_mark_block_used(int block, bool used)
+{
+	__ASSERT((block >= 0) && (block < NUM_BLOCKS), "block %d out of range", block);
+	pool.block_used[block] = used;
+	LOG_DBG("mark idx:%d = %u", block, used);
+}
+
+void npgps_print_blocks(void)
+{
+	char map[NUM_BLOCKS + 1];
+	int i;
+
+	LOG_INF("num blocks:%u, size:%u, first_free:%d", NUM_BLOCKS,
+		BLOCK_SIZE, pool.first_free);
+	for (i = 0; i < NUM_BLOCKS; i++) {
+		map[i] = pool.block_used[i] ? '1' : '0';
+	}
+	map[i] = '\0';
+	LOG_INF("map:%s", map);
+}
+
+
+int npgps_num_free(void)
+{
+	int num = 0;
+
+	for (int i = 0; i < NUM_BLOCKS; i++) {
+		if (!pool.block_used[i]) {
+			num++;
+		}
+	}
+	return num;
+}
+
+int npgps_find_first_free(int from_block)
+{
+	int i;
+	int start = from_block;
+
+	pool.first_free = NO_BLOCK;
+	for (i = 0; i < NUM_BLOCKS; i++) {
+		if (!pool.block_used[from_block]) {
+			pool.first_free = from_block;
+			break;
+		}
+		from_block = (from_block + 1) % NUM_BLOCKS;
+	}
+	LOG_DBG("first_free idx:%d from:%d", from_block, start);
+	return pool.first_free;
+}
+
+int npgps_offset_to_block(uint32_t offset)
+{
+	int block = offset / BLOCK_SIZE;
+
+	if ((block < 0) || (block >= NUM_BLOCKS)) {
+		LOG_ERR("invalid offset:%u", offset);
+		block = NO_BLOCK;
+	}
+	return block;
+}
+
+uint32_t npgps_block_to_offset(int block)
+{
+	if ((block < 0) || (block >= NUM_BLOCKS)) {
+		LOG_ERR("invalid block:%d", block);
+		return 0;
+	}
+	return block * BLOCK_SIZE;
+}
+
+int npgps_pointer_to_block(uint8_t *p)
+{
+	int ret = (uint32_t)(p - STORAGE_BASE) / BLOCK_SIZE;
+
+	LOG_DBG("ptr:%p to block idx:%d", p, ret);
+	if ((ret < 0) || (ret >= NUM_BLOCKS)) {
+		return NO_BLOCK;
+	}
+	return ret;
+}
+
+void *npgps_block_to_pointer(int block)
+{
+	void *ret;
+
+	if ((block < 0) || (block >= NUM_BLOCKS)) {
+		LOG_ERR("invalid block:%d", block);
+		ret = NULL;
+	} else {
+		ret = (void *)(STORAGE_BASE + block * BLOCK_SIZE);
+	}
+	LOG_DBG("block idx:%d to ptr:%p", block, ret);
+	return ret;
+}
+
+int npgps_download_init(npgps_buffer_handler_t handler)
+{
+	__ASSERT(handler != NULL, "must specify handler");
+	buffer_handler = handler;
+
+	return download_client_init(&dlc, download_client_callback);
+}
+
+int npgps_download_start(const char *host, const char *file, int sec_tag,
+			 const char *apn, size_t fragment_size)
+{
+	if (host == NULL || file == NULL) {
+		return -EINVAL;
+	}
+
+	int err;
+
+	err = k_sem_take(&pgps_active, K_NO_WAIT);
+	if (err) {
+		LOG_ERR("PGPS download already active.");
+		return err;
+	}
+	LOG_DBG("pgps_active LOCKED");
+
+	socket_retries_left = CONFIG_FOTA_SOCKET_RETRIES;
+
+	struct download_client_cfg config = {
+		.sec_tag = sec_tag,
+		.apn = apn,
+		.frag_size_override = fragment_size,
+		.set_tls_hostname = (sec_tag != -1),
+	};
+
+	err = download_client_connect(&dlc, host, &config);
```

Comment:
the `host` and `file` strings are on the stack and download client doesn't copy them... it looks like this will be a problem if the download client attempts to do a reconnect.

Reply:
they're pointers to a static copy in the calling function, but sure, I could do the strdup thing you do in FOTA.  I'll think about it.

Reply:
Ah I didn’t notice they were static 🤦